src/components/Menus/Menu.js

```diff
@@ -0,0 +1,204 @@
+    /** Callback function fired when a Menu is blurred */
+    onBlur: PropTypes.func,
+    /** Role HTML attribute */
+    role: PropTypes.string,
+    /** Customize style of menu parent */
+    style: PropTypes.shape({})
+  }
+
+  static defaultProps = {
+    role: 'menu',
+    maxHeight: 500,
+    ariaHidden: false,
+    style: {},
+  }
+
+  state = {
+    currentTabIndex: null
+  }
+
+  _handleClick = (option, index) => {
+    const { onSelect } = this.props
+
+    onSelect && onSelect(option, index)
+  }
+
+  _handleBlur = (event) => {
+    const { onBlur } = this.props
+    const currentTarget = event.currentTarget
+
+    setTimeout( () => {
+      if (!currentTarget.contains(document.activeElement)) {
+        this.setState({currentTabIndex: null}, () => {
+          onBlur && onBlur(event)
+        })
+      }
+    }, 0)
+  }
+
+  _handleKeyDown = (event) => {
+    const { onKeyDown } = this.props
+
+    switch(event.key) {
+      case 'ArrowDown':
+        event.preventDefault()
+        this.incrementTabIndex()
+        break
+      case 'ArrowUp':
+        event.preventDefault()
+        this.decrementTabIndex()
+        break
+    }
+
+    onKeyDown && onKeyDown(event)
+  }
+
+  _handleMenuItemFocus = (index) => {
+    this.setState({currentTabIndex: index})
+  }
+
+  blur() {
+    this.menu.blur()
+  }
+
+  focus() {
+    this.menu.focus()
+  }
+
+  incrementTabIndex() {
+    this.setState({currentTabIndex: this.nextValidTabIndex()})
+  }
+
+  decrementTabIndex() {
+    this.setState({currentTabIndex: this.prevValidTabIndex()})
+  }
+
+  nextValidTabIndex() {
+    const { currentTabIndex } = this.state
+    const menuItemChildren = this.getMenuItemChildren()
+    const maxIndex = menuItemChildren.length - 1
+    const newIndex = currentTabIndex !== null ? currentTabIndex+1 : 0
+
+    for (let index = newIndex; index <= maxIndex; index++) {
+      const menuItem = menuItemChildren[index]
+
+      if (this.menuItemIsValid(menuItem)) {
+        return index
+      }
+    }
+
+    return currentTabIndex
+  }
+
+  prevValidTabIndex() {
+    const { currentTabIndex } = this.state
+    const menuItemChildren = this.getMenuItemChildren()
+
+    for (let index = (currentTabIndex-1); index >= 0; index--) {
+      const menuItem = menuItemChildren[index]
+
+      if (this.menuItemIsValid(menuItem)) {
+        return index
+      }
+    }
+
+    return currentTabIndex
+  }
+
+  menuItemIsValid(menuItem) {
+    return menuItem && !menuItem.props.disabled
+  }
+
+  getMenuItemChildren() {
+    const { children } = this.props
+    const menuItemChildren = []
+
+    React.Children.map(children, (child) => {
+      if (child.type && child.type.displayName === 'MenuItem') {
+        menuItemChildren.push(child)
+      }
+    })
+
+    return menuItemChildren
+  }
+
+  renderChildren() {
+    const { children } = this.props
+    const { currentTabIndex } = this.state
+    let index = 0
+
+    return React.Children.map(children, (child) => {
```

Comment:
Kinda feels like `children` is the wrong way to pass the necessary information into this component. Since the only really "valid" children are `MenuItems` we should pass probably pass down `menuItems` as an array of hashes that `Menu` will consume and render. One of the configs for a given `menuItem` could be `divider: true` which would append a `Divider`.

With this pattern you don't have to do all the gymnastics of cloning the children, checking their validity, or trying to do complicated logic to increment the next selected item.

This way you can also more easily validate the `menuItems` passed down to ensure they are in the shape you expect.

Reply:
Even if the children were passed as a props they would still need to be cloned to pass `onClick` listener so the Menu can can respond to the onClick callback and do it's thing. I don't see how this will change whether it is children or props. It essentially doing the same thing just passed to Menu differently.

Reply:
You wouldn't pass the children as react components, you would pass them as native objects:
```
menuItems = [{label: 'United States', value:'US', divider: true}, {label: 'Canada', value: 'CA' }]
```

I made the arguments above, but basically by passing the data down this way will let you  operate on the data more easily.

Reply:
I thought of doing it that way, however, you lose the flexibility that `MenuItem` provides. It isn't just a simple component, it also allows for `children` that would render a completely custom component while keeping the `MenuItem` logic. 

Also, `MenuItem's` aren't the only valid children, they are the only valid children that `Menu`  will attach listeners to. For component library i'd prefer to error on the side of more flexibility rather than less.

Reply:
Ahh I see, okay nice 👍